wagtail/admin/static_src/wagtailadmin/js/core.js

```diff
@@ -538,6 +539,69 @@ wagtail = (function(document, window, wagtail) {
     $(document).ready(initDropDowns);
     wagtail.ui.initDropDowns = initDropDowns;
     wagtail.ui.DropDownController = DropDownController;
+
+    /**
+     * getTitleFromFilename expects a string as a result or any null/undefined if
+     * the value is not meant to be updated at all.
+     * @param {string=} filename 
+     * @param {Object} options 
+     * @param {string=} options.currentTitle - current entered value (if available), will be undefined for `ADD_MULTIPLE`
+     * @param {Number=} options.maxLength - maximum title length (if available), will be undefined for `ADD_MULTIPLE`
+     * @param {string} options.model - `DOCUMENT` or `IMAGE`
+     * @param {string} options.widget - `CHOOSER_MODAL`, `ADD` or `ADD_MULTIPLE`
+     * @returns {string=} if a string is not returned the title field will not be updated
+     */
+    function getTitleFromFilename (filename, options) {
+        if (options.currentTitle) return; // do not overwrite an existing entered title value
+        return filename.replace(/\.[^/.]+$/, ''); // remove file extension
+    }
+
+    wagtail.utils.getTitleFromFilename = getTitleFromFilename;
+
+    /**
+     * Returns a function to handle the change of a file (single or multiple) upload
+     * so that the context's title field can be updated based on the filename.
+     * @param {string} model - `DOCUMENT` or `IMAGE`
+     * @param {string} widget - `CHOOSER_MODAL`, `ADD` or `ADD_MULTIPLE`
+     * @returns {Function}
+     */
+    function getPopulateTitleHandler (model, widget) {
+        return function onChangeHandler (event, data) {
+
+            var $titleField;
+            var currentTitle;
+            var filename;
+
+            if (widget === 'ADD_MULTIPLE') {
```

Comment:
The branch here makes me think that there's not really much code in common between the ADD_MULTIPLE handler and the others, and we'd probably be better off making this into two separate functions.

Reply:
@gasman  I gave this a ago and personally I feel like it ends up being harder to understand, the reason is that the `onChangeHandler` needs to access the `model` & `widget` info to add to the `getTitleFromFilename` call. Splitting this two two functions ends up needing two more higher order functions which I think could create even more confusion. 

I tried a different approach, splitting the 'data getting' functions (eg. just make two small functions that each return `$titleField` & `currentTitle`). But this seems to make it confusing in a different way, plus we need to then either bind the function or pass in the `this` somehow so that the field can be used in single upload scenarios.

I have tried to isolate the 'pure data' handing into `getTitleFromFilename` and this function is what is left to smooth over the differences between the two upload widgets' internal data representation.

Plus the 'clean up' code at the end is partially shared between the various scenarios and splitting this could mean some duplication of this code that can easily get missed in the future.

Reply:
The problem here is that we have a centralised function in wagtailadmin's core.js, tasked with knowing the internal workings of 2-3 different form implementations from wagtailimages/wagtaildocs, so that it can extract values to plug into `getTitleFromFilename`. Those mechanisms have pretty much nothing in common - one obtains the title field from `event`, another obtains it from `data`; one gives us the filename in a clean format, the other makes us extract it from a fake path. This doesn't seem like a helpful division of responsibilities between wagtailadmin and wagtailimages/wagtaildocs - code that needs to know about the internals of an upload form should live somewhere close to that form (such as in add-multiple.js).

A useful exercise when working with the wagtailimages / wagtaildocs apps is to ask "what would a third-party app do if it wanted to do the same thing"? ([wagtailmedia](https://github.com/torchbox/wagtailmedia) is the canonical example of such an app.) As a general rule, we don't want to give wagtailimages / wagtaildocs any special treatment within wagtailadmin. `getTitleFromFilename` is mostly fine in that respect - while technically the docstring specifies that `model` has to be 'DOCUMENT' or 'IMAGE', it probably won't do any harm for a third-party app to call it and identify itself as the 'ADD_MULTIPLE' widget for the 'VIDEO' model. On the other hand, that third-party app couldn't safely call `getPopulateTitleHandler` with the same parameters, as there's no guarantee that its own multiple-uploader form has the same internals as the image/document ones. (Similarly, what if a third-party app added a new kind of upload widget, such as 'import from Google Drive', and wanted to piggyback on the get-title-from-filename logic? In this case, it couldn't make use of `getPopulateTitleHandler`, as it has no knowledge of what that form looks like.)

FWIW, the duplication of code between wagtailimages and wagtaildocs does bother me too, and I'd definitely like to see the multiple uploaders refactored into a shared generic implementation within wagtailadmin. But A) that's very much out of scope for this PR, and B) if that does happen, I'd still want to stick to the rule of "JS code that knows about form internals should live somewhere close to that form".

Reply:
Hey Matt. Thanks for explaining it - I see now what you mean about splitting the utils. Makes sense.

I'll give another crack at this when I get the chance and flag for a re-review then.